### ISSUE TITLE
iOS reinitialize tasks

### ISSUE THREAD

**Issue by VitaliiLytvynov** [SimonSimCity/Xamarin-CrossDownloadManager#32]
Can not reinitialize tasks that were started before the app was terminated or suspended.
And if I try to start download these urls again, app will crash.

**Comment by SimonSimCity** [comment]
@VitaliiLytvynov Can you please send me a crash-report next time you see it failing?

I can't reproduce it ... it works for me ...

**Comment by VitaliiLytvynov** [comment]
This is what I have in application output.

```
2016-10-27 12:39:03.552 MmvMMTravelIOS[3746:3540685] 
Unhandled Exception:
System.InvalidCastException: Specified cast is not valid.
2016-10-27 12:39:03.553 MmvMMTravelIOS[3746:3540685] Unhandled managed exception:
Specified cast is not valid. (System.InvalidCastException)
(null)
2016-10-27 12:39:03.553 MmvMMTravelIOS[3746:3540685] critical: Stacktrace:

2016-10-27 12:39:03.553 MmvMMTravelIOS[3746:3540685] critical: 
Native stacktrace:

2016-10-27 12:39:03.555 MmvMMTravelIOS[3746:3540685] critical: 	0   MmvMMTravelIOS                      0x001a0c5d mono_handle_native_sigsegv + 317
2016-10-27 12:39:03.555 MmvMMTravelIOS[3746:3540685] critical: 	1   MmvMMTravelIOS                      0x001aaa61 sigabrt_signal_handler + 161
2016-10-27 12:39:03.555 MmvMMTravelIOS[3746:3540685] critical: 	2   libsystem_platform.dylib            0x0b2dc79b _sigtramp + 43
2016-10-27 12:39:03.555 MmvMMTravelIOS[3746:3540685] critical: 	3   ???                                 0xffffffff 0x0 + 4294967295
2016-10-27 12:39:03.555 MmvMMTravelIOS[3746:3540685] critical: 	4   libsystem_c.dylib                   0x0af7c137 abort + 141
2016-10-27 12:39:03.556 MmvMMTravelIOS[3746:3540685] critical: 	5   MmvMMTravelIOS                      0x00365630 _ZL20print_all_exceptionsP10MonoObject + 0
2016-10-27 12:39:03.556 MmvMMTravelIOS[3746:3540685] critical: 	6   MmvMMTravelIOS                      0x001a1590 mono_invoke_unhandled_exception_hook + 96
2016-10-27 12:39:03.556 MmvMMTravelIOS[3746:3540685] critical: 	7   MmvMMTravelIOS                      0x001a033b mono_handle_exception_internal + 5611
2016-10-27 12:39:03.556 MmvMMTravelIOS[3746:3540685] critical: 	8   MmvMMTravelIOS                      0x0019ed49 mono_handle_exception + 41
2016-10-27 12:39:03.556 MmvMMTravelIOS[3746:3540685] critical: 	9   MmvMMTravelIOS                      0x001346c6 mono_x86_throw_exception + 182
2016-10-27 12:39:03.556 MmvMMTravelIOS[3746:3540685] critical: 	10  ???                                 0x1ddc5b57 0x0 + 500980567
2016-10-27 12:39:03.557 MmvMMTravelIOS[3746:3540685] critical: 	11  MmvMMTravelIOS                      0x003651bf xamarin_process_managed_exception_gchandle + 63
2016-10-27 12:39:03.557 MmvMMTravelIOS[3746:3540685] critical: 	12  MmvMMTravelIOS                      0x00365177 xamarin_ftnptr_exception_handler + 23
2016-10-27 12:39:03.557 MmvMMTravelIOS[3746:3540685] critical: 	13  ???                                 0x249a2630 0x0 + 614082096
2016-10-27 12:39:03.557 MmvMMTravelIOS[3746:3540685] critical: 	14  CFNetwork                           0x025fbac2 __46-[NSURLSession getTasksWithCompletionHandler:]_block_invoke_2 + 231
2016-10-27 12:39:03.557 MmvMMTravelIOS[3746:3540685] critical: 	15  CFNetwork                           0x0256bc34 __67-[__NSURLBackgroundSession _onqueue_getTasksWithCompletionHandler:]_block_invoke + 33
2016-10-27 12:39:03.557 MmvMMTravelIOS[3746:3540685] critical: 	16  Foundation                          0x0142b90b __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 12
2016-10-27 12:39:03.558 MmvMMTravelIOS[3746:3540685] critical: 	17  Foundation                          0x0142b5de -[NSBlockOperation main] + 105
2016-10-27 12:39:03.558 MmvMMTravelIOS[3746:3540685] critical: 	18  Foundation                          0x01429db6 -[__NSOperationInternal _start:] + 672
2016-10-27 12:39:03.558 MmvMMTravelIOS[3746:3540685] critical: 	19  Foundation                          0x01429b0e -[NSOperation start] + 70
2016-10-27 12:39:03.558 MmvMMTravelIOS[3746:3540685] critical: 	20  Foundation                          0x01425bbe __NSOQSchedule_f + 220
2016-10-27 12:39:03.558 MmvMMTravelIOS[3746:3540685] critical: 	21  libdispatch.dylib                   0x0ae9d6ef _dispatch_client_callout + 14
2016-10-27 12:39:03.558 MmvMMTravelIOS[3746:3540685] critical: 	22  libdispatch.dylib                   0x0ae81f3a _dispatch_queue_serial_drain + 908
2016-10-27 12:39:03.559 MmvMMTravelIOS[3746:3540685] critical: 	23  libdispatch.dylib                   0x0ae82728 _dispatch_queue_invoke + 1103
2016-10-27 12:39:03.559 MmvMMTravelIOS[3746:3540685] critical: 	24  libdispatch.dylib                   0x0ae844bc _dispatch_root_queue_drain + 384
2016-10-27 12:39:03.559 MmvMMTravelIOS[3746:3540685] critical: 	25  libdispatch.dylib                   0x0ae842d4 _dispatch_worker_thread3 + 134
2016-10-27 12:39:03.559 MmvMMTravelIOS[3746:3540685] critical: 	26  libsystem_pthread.dylib             0x0b2ec25c _pthread_wqthread + 1050
2016-10-27 12:39:03.559 MmvMMTravelIOS[3746:3540685] critical: 	27  libsystem_pthread.dylib             0x0b2e9f56 start_wqthread + 34
2016-10-27 12:39:03.559 MmvMMTravelIOS[3746:3540685] critical: 
=================================================================
Got a SIGABRT while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries 
used by your application.
=================================================================
```


Actually it happens when GetTask2() in DownloadManagerImplementation calls.

If I set there a breakpoint on this line:

`Queue.Add(new DownloadFileImplementation((NSUrlSessionDownloadTask)task));`

I receive this message in application output:

`MmvMMTravelIOS[3769:3577170] A background URLSession with identifier de.michael-mueller-verlag.MMTravel.BackgroundTransferSession already exists!`

**Comment by SimonSimCity** [comment]
Ok, this stacktrace tells me something about `Specified cast is not valid.` and I guess I know where it occurs.

Can you checkout master and see if the system now works as you expected? As soon as you've approved it, I'll publish a new version to NuGet.

**Comment by VitaliiLytvynov** [comment]
Nice.
It works.

Thank you for a fast response.

**Comment by SimonSimCity** [comment]
@VitaliiLytvynov Can you check if the downloads, that were running when the app was suspended, are resumed?

**Comment by VitaliiLytvynov** [comment]
Yes it works.

**Comment by SimonSimCity** [comment]
Just released 0.9.1 fixing this bug.
